### ISSUE TITLE
Non-C locale detected error

### ISSUE THREAD

**Issue by AndroidDesigner** [media-kit/media-kit#831]
Hello everyone.
I'm using this package in a ``flutter-pi`` project running on a ``raspberry-pi 4 bullseye`` to play and audio (mp3 or wav file).
But it crashes with the following error:
```
Non-C locale detected. This is not supported.
Call 'setlocale(LC_NUMERIC, "C");' in your code.
Segmentation fault
```

what is the problem and solution?
Thanks in advance!

**Comment by abdelaziz-mahdy** [comment]
searching thorough google and chatgpt that was gpt answer, since i cant test it, i posted it here since it may help you dont know if works or not


The error message indicates that the `flutter-pi` project is running into a locale issue. This happens because the audio playback library you're using requires the numeric locale to be set to "C". The segmentation fault (segfault) occurs due to this locale mismatch.

Here’s a step-by-step guide to resolve this issue:

### 1. Modify the Locale Setting in Your Code

You need to set the numeric locale to "C" at the start of your application. In a Flutter project, you can do this by calling native C code to set the locale.

**Steps:**
1. **Create a C/C++ file** (e.g., `locale_fix.c`):
    ```c
    #include <locale.h>

    void set_numeric_locale_to_C() {
        setlocale(LC_NUMERIC, "C");
    }
    ```

2. **Compile the C file** to a shared library. You can use GCC for this:
    ```sh
    gcc -shared -o liblocale_fix.so -fPIC locale_fix.c
    ```

3. **Load and Call the Shared Library in Dart**:
    - Add the `ffi` package to your `pubspec.yaml`:
        ```yaml
        dependencies:
          ffi: ^1.1.2
        ```
    - Create a Dart file to load and call the C function (e.g., `locale_fix.dart`):
        ```dart
        import 'dart:ffi';
        import 'dart:io';

        typedef SetNumericLocaleC = Void Function();
        typedef SetNumericLocaleDart = void Function();

        void setNumericLocaleToC() {
          final DynamicLibrary dylib = Platform.isLinux
              ? DynamicLibrary.open('liblocale_fix.so')
              : DynamicLibrary.process();

          final SetNumericLocaleDart setLocale = dylib
              .lookup<NativeFunction<SetNumericLocaleC>>('set_numeric_locale_to_C')
              .asFunction();

          setLocale();
        }
        ```

4. **Call the Locale Fix Function** at the start of your application (e.g., in `main.dart`):
    ```dart
    import 'locale_fix.dart';

    void main() {
      setNumericLocaleToC();
      // Your existing Flutter setup code
      runApp(MyApp());
    }
    ```

### 2. Rebuild and Run Your Flutter Application

After adding the locale fix, rebuild your application and run it again on the Raspberry Pi.

### Additional Considerations

- **Check Dependencies**: Ensure all your project dependencies are compatible with your environment and the Raspberry Pi architecture.
- **Debugging**: If the problem persists, use tools like `gdb` to get a more detailed stack trace of the segmentation fault, which can help identify the root cause more precisely.

By setting the numeric locale to "C", you align with the library's expectations, which should prevent the segmentation fault and allow your audio to play correctly.

**Comment by alexmercerind** [comment]
Yup... we have same call on GNU/Linux. Regardless, flutter-pi or any other embedder support is not there.

**Comment by Krakert** [comment]
@abdelaziz-mahdy I have tried to follow the code provided, but I was unable to get it to work. The error I get: `liblocale_fix.so: cannot open shared object file: No such file or directory`.

@alexmercerind is Flutter-pi support something that might be added in the future? I have tried the `flutterpi_gstreamer_video_player` package and it is given me trouble when playing video, GStreamer is not working correct on a Pi 4:
```
** (flutter-pi:3491): WARNING **: 13:48:00.613: v4l2h264dec0: 1 frames 677-677 left undrained after CMD_STOP, eos sent too early: bug in decoder -- please file a bug

(flutter-pi:3491): GStreamer-CRITICAL **: 13:48:00.660: gst_sample_set_caps: assertion 'gst_sample_is_writable (sample)' failed
```

I was hoping to use `media-kit` to try and run video's that way.

**Comment by alexmercerind** [comment]
@Krakert 

https://github.com/media-kit/media-kit/blob/88c49c6c0f6187a131a4053f0f1a3a42f2922a8a/libs/linux/media_kit_libs_linux/linux/media_kit_libs_linux_plugin.cc#L16

It's required on GNU/Linux.

<hr>

> Is Flutter-pi support something that might be added in the future?

The audio setup of media-kit works out of the box on every platform... that's due to project's architecture. The video rendering uses platform / embedder specific APIs for texture etc.

I don't really own a Raspberry Pi, neither I have any interest... contributions welcomed.

**Comment by Krakert** [comment]
So it should work, but where do i need to add the files and build the `liblocale_fix.so` because it is unable to find it.

**Comment by marinct** [comment]
You can try:
`sudo update-locale LC_NUMERIC=C`
